lib/api/ldp/ldp-request.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const HttpError = require('standard-http-error')
+
+class LdpRequest {
+  /**
+   * @param options {object}
+   *
+   * @param options.host {SolidHost} Host / config object
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.req {HttpRequest} Save the Express req object, useful
+   *   to access headers/preferences, writeStream for body, etc.
+   *
+   * @param [options.authn=null] {object} Authentication object, contains WebID
+   *   string, as well as any bearer credentials/tokens. Needed for
+   *   authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.parsedBody] {?} Request body, parsed when appropriate
+   *   (when content-type is parse-able)
+   */
+  constructor (options) {
+    this.host = options.host
+    this.target = options.target
+    this.req = options.req
+    this.authn = options.authn
+    this.parsedBody = options.parsedBody
+  }
+
+  static from ({req, host}) {
+    let request
+
+    switch (req.method) {
+      case 'head':
+        /**
+         * Checks to see if target exists
+         */
+        request = LdpHeadRequest.from({req, host})
+        break
+      case 'get':
+        /**
+         * Use cases to handle:
+         *  - LdpListContainerRequest, if target is an ldp container
+         *  - However, if Accept: header is HTML, check if index.html exists
```

Comment:
It would be nice to be able to dispatch between "List container" and implicit "I'm requesting a container url but really get me its index.html/index.ttl" type of GET requests here, during parsing & constructing the Request object.

But that requires passing in a backend object (e.g. `LdpFileStore`), to check if said `index.html` exists. Which I was initially reluctant to do, in terms of separation of concerns (and only pass in the backend into the `request.handle()` method). But it seems like it might be a good idea to pass it in at this early stage.

Reply:
Then maybe `target` should be parsed first before this decision happens. You should be able to use https://github.com/solid/node-solid-server/blob/v4.0.14/lib/resource-mapper.js for that.